Mage/src/main/java/mage/target/targetadjustment/EachOpponentPermanentTargetsAdjuster.java

```diff
@@ -0,0 +1,41 @@
+package mage.target.targetadjustment;
+
+import mage.abilities.Ability;
+import mage.filter.Filter;
+import mage.filter.predicate.permanent.ControllerIdPredicate;
+import mage.game.Game;
+import mage.game.permanent.Permanent;
+import mage.players.Player;
+import mage.target.TargetPermanent;
+
+import java.util.UUID;
+
+/**
+ *
+ * @author notgreat
+ */
+public class EachOpponentPermanentTargetsAdjuster implements TargetAdjuster {
```

Comment:
Current "class" logic without copyable support is buggy and will fail after any copy refactor/fix (classes without override copy methods will lose inner data on parent's copy call -- popular bug in old days).

If you want to introduce a class for adjuster (e.g. dynamic adjusters) then must do more refactors here or in another PR:
* TargetAdjuster must be Copyable due code style;
* All other enums must be migrated to Copy code too (override Copy method with instance return);
* All usage of TargetAdjuster in ability and other copy constructors must use adjuster.copy();
* xxxTargetsAdjuster must be renamed to TargetAdjuster due naming style;
* baseTarget must be renamed to blueprintTarget;
* must add some docs on usage like "Duplicate permanent target for each opponent. Warning, do not use controlled filters because it will be setup inside)"

I recommends to rework current implementation and keep old enum logic (static adjusters), but with additional inheritance. Example:

Base adjuster:
``` java
/**
 * Duplicate permanent target for each opponent.
 */
interface EachOpponentPermanentTargetAdjuster extends TargetAdjuster {

    /**
     * Warning, do not use controlled filters because it will be setup inside for each opponent
     */
    TargetPermanent prepareTarget(Ability ability, Game game);

    default void adjustTargets(Ability ability, Game game) {
        // real implementation here
    }
}
```

Inherited adjusters:
``` java
enum BlatantThieveryTargetAdjuster implements EachOpponentPermanentTargetAdjuster {
    instance;

    @Override
    public TargetPermanent prepareTarget(Ability ability, Game game) {
        return new TargetPermanent(0, 1, StaticFilters.FILTER_PERMANENT);
    }
}
```

Reply:
I don't think I understand the issue. Currently, the data stored in `TargetAdjuster` is set once during construction and then never modified. I could copy the target during construction if you would like me to make that fact a guarantee, though I can't think of a reasonable situation where the target would be modified.

Are you saying that since `TargetAdjuster` might at some point in the future become `Copyable` then we must not do this? It seems to me that it would be enough to add a comment to the `TargetAdjuster` class noting the requirement of not adding a copy method without also refactoring this code.

> xxxTargetsAdjuster must be renamed to TargetAdjuster due naming style;

I pluralized it based on `XTargetsAdjuster`, since both this and that adjuster will target multiple permanents whereas the other adjusters named `TargetAdjuster` all have exactly one target.

Reply:
I also don't understand why the current PR implementation would be considered buggy or why `Copyable` is necessary.

Reply:
I do agree about adding the doc comment not to use filter controlled or controller predicate in blueprint filter

Reply:
It’s a standard code style to protect from a very weird and hard finding bugs (data modification from another game/simulation/rollback/etc). All class objects must be copyable (deep copy), all static objects/enums/singleton/values can use ref copy.

If you introduce adjuster as dynamic object/class then all related code (copy constructors) must use it as copyable object (targetAdjuster = effect.targetAdjuster == null ? null : effect.targetAdjuster.copy()).

Reply:
That's only necessary if the class object is actually mutable (internal state changing on usage). It's not always the case that it's required. For example, another functional interface in XMage is `Condition`, which is not `Copyable`.